assets/src/models/ladderVehiclePosition.ts

```diff
@@ -0,0 +1,156 @@
+import { Vehicle } from "../skate"
+
+export enum VehicleDirection {
+  Up,
+  Down,
+}
+
+export interface LadderVehiclePosition {
+  vehicle: Vehicle
+  x: number
+  y: number
+  vehicleDirection: VehicleDirection
+  lane?: number
+}
+
+type Range = [number, number]
+
+export const vehicleDirectionOnLadder = (
+  vehicle: Vehicle,
+  ladderDirection: LadderDirection
+): VehicleDirection =>
+  (vehicle.direction_id === 1) ===
+  (ladderDirection === LadderDirection.ZeroToOne)
+    ? VehicleDirection.Down
+    : VehicleDirection.Up
+
+export const putIntoLanes = (
+  ladderVehiclePositions: LadderVehiclePosition[]
+): LadderVehiclePosition[] =>
+  ladderVehiclePositions
+    .slice()
+    .sort(byDirectionAndY)
+    .reduce(
+      (
+        previousLadderVehiclePositions: LadderVehiclePosition[],
+        ladderVehiclePosition: LadderVehiclePosition
+      ): LadderVehiclePosition[] => {
+        const overlappers = overlappingPreviousVehicles(
+          previousLadderVehiclePositions,
+          ladderVehiclePosition
+        )
+        const occupiedLanes: number[] = overlappers.reduce(
+          (lanes: number[], overlapper: LadderVehiclePosition) =>
+            overlapper.lane !== undefined ? [...lanes, overlapper.lane] : lanes,
+          []
+        )
+        const lane = firstOpenLane(occupiedLanes)
+
+        return [
+          {
+            ...ladderVehiclePosition,
+            lane,
+          },
+          ...previousLadderVehiclePositions,
+        ]
+      },
+      []
+    )
+
+export const xInLaneWithOffset = (ladderVehicleHorizontalOffset: number) => ({
+  x,
+  vehicleDirection,
+  lane,
+}: LadderVehiclePosition): number => {
+  const laneOffset = (lane || 0) * ladderVehicleHorizontalOffset
+  return vehicleDirection === VehicleDirection.Up
+    ? x + laneOffset
+    : x - laneOffset
+}
+
+export const byDirectionAndY = (
+  a: LadderVehiclePosition,
+  b: LadderVehiclePosition
+): number => {
+  if (a.vehicleDirection === VehicleDirection.Down) {
+    if (b.vehicleDirection === VehicleDirection.Up) {
+      return -1
+    } else {
+      if (a.y > b.y) {
+        return -1
+      } else if (b.y > a.y) {
+        return 1
+      } else {
+        return 0
+      }
+    }
+  } else {
+    if (b.vehicleDirection === VehicleDirection.Down) {
+      return 1
+    } else {
+      if (a.y < b.y) {
+        return -1
+      } else if (b.y < a.y) {
+        return 1
+      } else {
+        return 0
+      }
+    }
+  }
+}
+
+const overlappingPreviousVehicles = (
+  previousLadderVehiclePositions: LadderVehiclePosition[],
+  ladderVehiclePosition: LadderVehiclePosition
+): LadderVehiclePosition[] =>
+  previousLadderVehiclePositions.filter(previous =>
+    areOverlapping(previous, ladderVehiclePosition)
+  )
+
+export const areOverlapping = (
+  a: LadderVehiclePosition,
+  b: LadderVehiclePosition
+): boolean => {
+  return (
+    a.vehicleDirection === b.vehicleDirection &&
+    areOverlappingRanges(yRange(a), yRange(b))
```

Comment:
This could just be `abs(a.y - b.y) < minDistanceBetween`, and you wouldn't need the range stuff.

Reply:
Nice!